### ISSUE TITLE
request !  calibration

### ISSUE THREAD

**Issue by nesquik011** [cncjs/cncjs#376]
i request calibration for all axis , manually press bottom , and once the server runs it auto calibration all supported axis please and option to disable the auto function thank you a lot

**Comment by justinclift** [comment]
This is a bit confusing.  When you press the button, what do you want it to do instead? :smile:

**Comment by nesquik011** [comment]
ok its like every axis , hit the limit switch to know where it is then go to a point XYZABC i write down maybe i will pick home at zero   ,  or maybe i want them all to hit their limit switches and know they location and stay there waiting for Gcode

**Comment by nesquik011** [comment]
once the Gcode ends everything go to that location i set the calibration at

**Comment by justinclift** [comment]
Ahhh.  It sounds like you want a macro button maybe? I think CNCjs has macro buttons, though I personally haven't tried that yet. :smile:

**Comment by nesquik011** [comment]
maybe it sound like macro ,  but i dont know ,   but what i know its amazing , i saw a lot of machine do that function ,  if axis was on   3,4,5   once you turn on the machine they move to hit the limit switch and stay there or go to center , look amazing , and once its done it go back to hit limit switch and go the place where someone already set it

**Comment by TobyGWilliams** [comment]
We have this already on our CNC machines running cncjs.

Set your home position to be different to your limit switch locations (not a cncjs thing). When our machines home, they set the homing position to be -2, -2, -2, this means that when you say go to zero, the machine wont quite reach the limit switch positions.

If you do this, you have to be aware that **homing** and **go to home** are different.

Toby

**Comment by MitchBradley** [comment]
It seems like you want a professional machine at a DIY price.  Pro machines are build around known sets of components and are programmed by people who are paid a lot to eliminate second-order and third-order annoyances.  The manufacturers charge real money for that level of system integration.

**Comment by nesquik011** [comment]
Hmm i dont think its hard option to add , its like a macro , turn the machine on ? Then boom xyzabc all move to limit switch hit then go back to some fixed coordination we would manukky write and so on after finishing the milling do the same

**Comment by MitchBradley** [comment]
The macro is easy; I have essentially that Gcode in a setup section of the milling program that I use the most frequently.  It does a homing cycle to XYZ, sets the machine coordinates to the limit values, moves to a fixed offset, sets the XY work coordinates to 0, then does a probe cycle to set Z.  That is easy, it is just GCode.  I have a version of it in a separate file so if I want to run it at anytime I just call up that file and say go.  You could store it as a macro instead of in a file.
But you seem to want deep integration with the machine so it happens automatically when you turn "the machine" on.  What is "the machine"?  Is it the stepper drivers, or the microprocessor running Grbl/TinyG/g2core/Marlin/smoothieware, or the computer running the cncjs server code, or the user interface computer that is running the browser?  What if all of those things have separate switches?   How does the cncjs server computer find out that the microprocessor has been turned on?  Is the method of finding out different for grbl vs TinyG etc?  (it almost certainly is).  What if the user interface computer has not yet connected?
cncjs is a component that can be used for many different kinds of machines, using many different combinations of computers, communicating over different kinds of interconnecting links, running different kinds and versions of machine controller software.
What you want would might be relatively easy to do for a specific machine if you know what all the components are and they are all turned on by the same switch and you can guarantee that they finish booting in a specific order.  But making it generally useful, so that it works for more than one person's unique machine configuration, would be very difficult.  You would pretty much have to do all the work again, from scratch, for every different machine.

**Comment by justinclift** [comment]
@nesquik011 As a thought, are you ok with learning how to program yet or are you still not interested in that? :smile:

Asking because the stuff you're thinking of can probably be done, but you'll need to be the person who makes it happen.  If you're ok to learn the programming needed, that would be useful. :smile:

**Comment by nesquik011** [comment]
@justinclift  if you know a place online teach how to make controller like cncjs , i will join and learn , i already searched but i didnt find something like that , and i did try to generally understand javascript it kinda not welcoming me :( but i want to learn  , i would keep the updates going on

**Comment by nesquik011** [comment]
@MitchBradley ok cncjs itself the software on pi or pc or mac   just a new wed jet  i will try to make a image for that

**Comment by justinclift** [comment]
@nesquik011 Cool. :smile:

There are many places around for learning Javascript.  Here's one from a random search, and they have many courses:

&nbsp; &nbsp; https://www.coursera.org/courses?query=javascript

I've also personally come across some courses that didn't work well (for me) with other programming languages.  When that happens, I find it best to do something else for a short time (say 15 minutes) to clear the head, then go and find a different course + try again.  It can sometimes take a few attempts, but they do seem to add up and it works in the end.  Persistence in learning does get there in the end. :smile:

**Comment by nesquik011** [comment]
@justinclift  thats amazing i will try with that again :D

**Comment by nesquik011** [comment]
@MitchBradley   @cheton          [https://ibb.co/gZuYOK](url)
                                                     
                                                     [https://ibb.co/jjcRce](url)   

please give my idea a chance ,dont just kill it by says its hard its amazing and as i need it i am sure others need it as well , and i see its just simple as macro maybe we call it brain ? as mach3 lol anyway  please see the pictures i uploaded two pix

**Comment by MitchBradley** [comment]
Here is the macro for performing a homing cycle (g28) then setting the work coordinate system origin (g10) relative to the homing position.  In this example the work origin is 30 mm to the left of the X limit, 40 mm in from the Y limit, and 50 mm down from the Z limit.  The first line 'g21 g90' just makes sure that the system is in mm mode and absolute mode, in case you run the macro when the system is in some other state.

```
g21 g90
g28.2 x0 y0 z0
g10 l2 p0 x-30 y-40 z-50
```
cncjs already lets you set up a macro like this and run it at the push of a button.

You don't seem to believe me when I tell you that automating this based on state changes of system sub-components is really tricky, depending on many factors that are difficult to control in a cncjs system setup.  Perhaps, after you learn programming, and try implementing it yourself, and live with all the corner cases, you might understand what I am saying.  I am a programmer and electrical engineer by trade, with a specialty in computer system startup software, so I have spend an enormous amount of time, spanning 40 years, dealing with the coordination of multiple system components as they start.  I am also a daily user of a cncjs-based milling setup on which my current business depends.  But you don't have to believe me; try automating it yourself.

Or just use the existing cncjs macro facility, plug in the numbers for your machine in the macro above, and punch the button when you want to rezero the machine.

**Comment by nesquik011** [comment]
ok how to make macro auto run ? once its connected to the controller ?  if that option is not there then may you add it so i can write the macro to do that function automatically , and i believe you , its just a request not just personally everyone would use it as well , i want cncjs to at be the best form of it

**Comment by justinclift** [comment]
@nesquik011 You're probably best off to start that javascript learning again, so you can add the bits yourself.

It will probably take you a few weeks until you have a decent grasp of doing stuff with javascript, then you should be ok to start trying to modify cncjs.

From my point of view it's a bit frustrating to see you spend **months** of your own time asking people to do stuff for you, when I know you could learn the bits you want and then add this stuff yourself.  And share it with everyone.

That's why I keep encouraging you to get back into the javascript learning.  You really can get this stuff done. :smile:

**Comment by nesquik011** [comment]
@justinclift you are right , i searched online there is something called internet of thing i think cncjs is one but not sure , so what do i do ? just learn basic java script

**Comment by justinclift** [comment]
Yep.  Start with learning basic javascript.

But, try some of those Coursera courses.  They should do more than just the very basic bits, and progress into more powerful stuff.  That will be the good stuff you want. :smile:

**Comment by MitchBradley** [comment]
The first time that it auto-runs by mistake and breaks something, because the auto-run logic was faulty (and it will be faulty until you have discovered and fixed several situations that are hard to predict), you may change your mind about having it auto-run.

cncjs is only superficially written in JavaScript.  To understand it you need to become familiar with a lot of frameworks that are on top of (or underneath) JavaScript, such as React, bootstrap, Node, and that thing that translates the new version of ECMAscript into JavaScript.

**Comment by nesquik011** [comment]
I believe nothing will break there is end stop , its nt a big deal to make the axis hit and end stop then go to some location , i dont know why you believe its a bad idea ? I want that function so lets wait to Cheton or any other interested developer to decided will they make it or no

**Comment by nesquik011** [comment]
@cheton  hello may you read my request please
